### PR TITLE
Initial conversion of the specification to bikeshed flavoured Markdown

### PR DESCRIPTION
**Pull request by joshuagl**
This PR lays the groundwork for #121 and is based on the excellent initial work by @erickt.

You can see a rendered version of the bikesheded specification document here: https://joshuagl.github.io/specification/

The major changes in this PR are:
* Convert tuf-spec.md to Bikeshed Flavoured Markdown (bfm) in 2be515a
* Start adopting some of the bfm features in 0713382 and c714c04. __NOTE__ until this point the content of the spec is unchanged, all of the changes are stylistic
* Changes the wording of some of the object field definitions, to better match the new layout and style, in 211ffbf
* Misc. tweaks in remaining commmits

I've been pleasantly surprised by how readable the [raw document](https://raw.githubusercontent.com/joshuagl/specification/joshuagl/bikeshed-initial/tuf-spec.md) still is, and how nice the rendered version looks. The styling of code blocks, cross-linking of definitions, and hyper-linked contents are a major improvement to the readability of the rendered specification.

Because the raw version isn't significantly more complex to edit, but handling additional changes to the specification makes maintaining this in a branch awkward, I would like to propose that we consider merging this PR and continuing to work on next steps (see below) with additional PRs against the main branch.

cc @JustinCappos @trishankatdatadog @mnm678 @lukpueh 

Next steps:
* Agree whether to proceed with switching to Bikeshed flavoured Markdown & merge this PR
* Define a CI pipeline that will
  * automatically publish updated versions of the specification (master branch) to i.e theupdateframework.github.io/specification
  * preserve, in a versioned location, a published version for each tagged release of the specification to i.e. theupdateframework.github.io/specification/1.0.16/
  * automatically publish updated version of the draft specification (draft branch) to i.e. theupdateframework.github.io/specification/draft
* Convert the detailed client workflow to call out to subsections (see #121)

### CODE REVIEW

**Comment by erickt** [comment]
This is great, thanks @joshuagl! One thing I didn't get a chance to explore is that we could leverage a github action like https://github.com/koumoul-dev/gh-pages-multi to publish the main branch to a `/draft` directory, then have a separate branch tracks the latest stable version publish to a separate directory.

**Review by mnm678 — DISMISSED** [review]
This is a big improvement for readability. The links within the document and overall format make the rendered version much easier to reference. I support merging this as is, and adding automation later.

**Comment by trishankatdatadog** [comment]
ZOMG, this is fantastic, thanks @joshuagl and @erickt 👏🏽 

QQ: the main spec version is 1.0.17, but there are places throughout the spec where we say 1.0.0. Should they all say 1.0.17?

**Comment by lukpueh** [comment]
Amazing work, @joshuagl and @erickt! And good question, @trishankatdatadog. Based on a quick grep, we say "1.0.0"  in the section about supported taps and in example metadata (`spec_version` field). I think we could change the former to say "1.y.z", i.e. "this major version" instead of "this version".
The latter I'd probably leave as is and not bump with every (patch) release, because they are "only" examples and adopters are free to interpret them as they wish, also they must be numeric, so I'd rather not use placeholders in the example.

**Comment by joshuagl** [comment]
> This is great, thanks @joshuagl! One thing I didn't get a chance to explore is that we could leverage a github action like https://github.com/koumoul-dev/gh-pages-multi to publish the main branch to a `/draft` directory, then have a separate branch tracks the latest stable version publish to a separate directory.

Nice find. This exactly the kind of workflow I was planning to create, I will try and experiment with gh-pages-multi this week. Thanks @erickt

**Comment by joshuagl** [comment]
> Amazing work, @joshuagl and @erickt! And good question, @trishankatdatadog. Based on a quick grep, we say "1.0.0" in the section about supported taps and in example metadata (`spec_version` field). I think we could change the former to say "1.y.z", i.e. "this major version" instead of "this version".
> The latter I'd probably leave as is and not bump with every (patch) release, because they are "only" examples and adopters are free to interpret them as they wish, also they must be numeric, so I'd rather not use placeholders in the example.

I agree with the rationale here, I've pushed a change to say "this major version (1.x.y) of the specification" in f4b15a9

**Comment by joshuagl** [comment]
I've pushed a few additional tweaks since the initial comments. I started working on a publishing pipeline using GitHub Actions, but didn't get it "done enough" to include here before heading out over the winter break.

I will pick up the automation piece of this in the new year.

**Comment by joshuagl** [comment]
Automating this with Github Actions pushing to GitHub Pages is a bit clunky, especially as I'm trying to avoid using 3rd-party actions.

My current proposal is:
1. a CI workflow will build each PR and publish to an appropriately named directory in the `gh-pages` branch, i.e. this PR might be published to `wip/pulls/143` (and then would be browsable at theupdateframework.github.io/specifcation/wip/pulls/143). The workflow would also update the PR with a link to the rendered document.
2. a cleanup workflow would run periodically and compare the children of `wip/pulls` to open PRs, removing any children of `wip/pulls` for which there isn't a corresponding open PR
3. a release workflow will be triggered on changes to `master`, and:
    1. make a release on GitHub with a corresponding versioned tag
    2. build the spec for that release and publish it to the `latest` directory of the `gh-pages` branch
    3. also publish the built spec to a versioned subdirectory of the `gh-pages` branch
    4. update a listing page (index.html?) to point to the versioned subdirectory (the listing will enable browsers to find: the latest spec, specific versioned spec releases, and the most recent draft)
4. a draft workflow will be triggered on changes to `draft`, and:
    1. build the spec in the `draft` branch
    2. publish the spec to a `draft` subdirectory of the `gh-pages` branch

Questions:
* Should we be signing spec releases (and therefore not doing automated releases on push to master)?
* Do we want official (latest and versioned) releases of the spec to be published in a separate place to WIP (PRs) and draft releases? i.e. we could publish released versions of the spec on theupdateframwork.io via Netlify and only use GitHub Pages (theupdateframework.github.io/specification) to host WIP (draft & PR) specification versions.
* Is the cleanup workflow necessary or, for the sake of a ~200K per PR, should we retain history for the PRs?
* Does everyone distrust 3rd party actions, or is that just me? If we're only using actions to automate WIP changes, and opt to manually make signed spec releases (and push them to theupdateframework.io), I would personally be less anxious about using 3rd party actions.

---

Having written this all up, I think I should probably create a new issue for the automation and handle it separately to this PR.

@JustinCappos, @trishankatdatadog, @mnm678, @lukpueh can we review this PR as is with a view to merging and handling automation later? @mnm678 [indicated](https://github.com/theupdateframework/specification/pull/143#pullrequestreview-550520536) she was OK with that approach (but my adding to the PR dismissed her review) , any objections?

**Comment by joshuagl** [comment]
Close/re-open to prod Travis.

**Comment by lukpueh** [comment]
> Automating this with Github Actions pushing to GitHub Pages is a bit clunky, especially as I'm trying to avoid using 3rd-party actions.
> 
> My current proposal is:
>  ...

This all sounds wonderful to me, and I'm looking forward to reviewing the GH Actions (and learn).

Have you also thought about auto-syncing master with draft on a release? I suppose this will usually involve manual conflict resolving, so maybe the action can at least create a corresponding issue on each release so that maintainers don't forget?


> Questions:
> 
> * Should we be signing spec releases (and therefore not doing automated releases on push to master)?

If we release manually, then yes. If we release automatically, then probably yes, although that signature seems less significant.

> * Do we want official (latest and versioned) releases of the spec to be published in a separate place to WIP (PRs) and draft releases? i.e. we could publish released versions of the spec on theupdateframwork.io via Netlify and only use GitHub Pages (theupdateframework.github.io/specification) to host WIP (draft & PR) specification versions.

I don't think it's worth the effort. Making `theupdateframework.io` point directly to latest and an index page that lists the versioned history should be just as good.

> * Is the cleanup workflow necessary or, for the sake of a ~200K per PR, should we retain history for the PRs?

I don't think the build history for all PRs is of much interest, but I also don't think an automated cleanup workflow is worth much effort, at least for now. Or does the workflow already exist, then we should use it. 

> * Does everyone distrust 3rd party actions, or is that just me? If we're only using actions to automate WIP changes, and opt to manually make signed spec releases (and push them to theupdateframework.io), I would personally be less anxious about using 3rd party actions.

Yes, I am not very happy about them either, and I did make some suggestion to mitigate this threat (https://github.com/theupdateframework/tuf/issues/1246#issuecomment-748045416). But I think I'll let others decide here.

*On a related side-note: It would be really cool if there were multiple independent builders / release taggers and a software supply chain framework to encode authorization and threshold requirements...* :P


At any rate, if we decide to keep signed releases a manual task, then I'd also love to have a GitHub action that auto-creates an issue to remind maintainers of that task.
> 
> Having written this all up, I think I should probably create a new issue for the automation and handle it separately to this PR.

👍 

> 
> @JustinCappos, @trishankatdatadog, @mnm678, @lukpueh can we review this PR as is with a view to merging and handling automation later? @mnm678 [indicated](https://github.com/theupdateframework/specification/pull/143#pullrequestreview-550520536) she was OK with that approach (but my adding to the PR dismissed her review) , any objections?

I'm fine with this. But I haven't looked at the diff yet. I can do that tomorrow and green-light this PR.

**Comment by trishankatdatadog** [comment]
> @JustinCappos, @trishankatdatadog, @mnm678, @lukpueh can we review this PR as is with a view to merging and handling automation later? @mnm678 [indicated](https://github.com/theupdateframework/specification/pull/143#pullrequestreview-550520536) she was OK with that approach (but my adding to the PR dismissed her review) , any objections?

Yes, this stuff is beautiful, let's go ahead. Someone else should double-check that we haven't lost anything in translation.

**Comment by joshuagl** [comment]
> > Automating this with Github Actions pushing to GitHub Pages is a bit clunky, especially as I'm trying to avoid using 3rd-party actions.
> > My current proposal is:
> > ...
> 
> This all sounds wonderful to me, and I'm looking forward to reviewing the GH Actions (and learn).
> 
> Have you also thought about auto-syncing master with draft on a release? I suppose this will usually involve manual conflict resolving, so maybe the action can at least create a corresponding issue on each release so that maintainers don't forget?

I dismissed the idea of auto-synching, because in my (limited) experience there's usually some manual conflict resolving. Filing an issue on merge is a great idea, though!

> > * Does everyone distrust 3rd party actions, or is that just me? If we're only using actions to automate WIP changes, and opt to manually make signed spec releases (and push them to theupdateframework.io), I would personally be less anxious about using 3rd party actions.
> 
> Yes, I am not very happy about them either, and I did make some suggestion to mitigate this threat ([theupdateframework/tuf#1246 (comment)](https://github.com/theupdateframework/tuf/issues/1246#issuecomment-748045416)). But I think I'll let others decide here.
> 
> _On a related side-note: It would be really cool if there were multiple independent builders / release taggers and a software supply chain framework to encode authorization and threshold requirements..._ :P

🤔 

> At any rate, if we decide to keep signed releases a manual task, then I'd also love to have a GitHub action that auto-creates an issue to remind maintainers of that task.

💯 this is a fine idea.

> > @JustinCappos, @trishankatdatadog, @mnm678, @lukpueh can we review this PR as is with a view to merging and handling automation later? @mnm678 [indicated](https://github.com/theupdateframework/specification/pull/143#pullrequestreview-550520536) she was OK with that approach (but my adding to the PR dismissed her review) , any objections?
> 
> I'm fine with this. But I haven't looked at the diff yet. I can do that tomorrow and green-light this PR.

Brilliant, thank you @lukpueh !

**Review by lukpueh — COMMENTED** [review]
Now that I've taken a closer look, this seems even more impressive. Thanks again, @joshuagl!

So far I've rather superficially compared the old rendered markdown with the new rendered html, which looks really good. I'd like to give it another more thorough (word-diff) pass, especially in regards to renumbering, before I green-light.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:19]: Oh, I already have changes for release_checker on a branch. I will find them and push to this PR

File: tuf-spec.md
Comment (lukpueh) [reply tuf-spec.md:1292]: I was thinking of the discussion in https://github.com/theupdateframework/specification/pull/133.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:19]: Pushed in 3aa6863

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:4]: Thanks for catching this, I had intended to raise this question myself. We could, probably should, use a different status when building the release versions of the specification.

Perhaps TS: "Technical Specification" for releases and DTS: "Draft Technical Specification" for non-releases?

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:1292]: I had originally used auto-numbering but switched to manually numbering steps in the client workflow in https://github.com/theupdateframework/specification/pull/143/commits/878f3cb4adec50855a7a6066e3ac9b6585fecd3a. I would like us to switch to auto-numbering, but for that to work we need to be able to cross-reference other steps without having to know what their step number is.

Comment (joshuagl) [reply tuf-spec.md:1292]: This is definitely something to capture as future work, if/when this PR lands.

File: tuf-spec.md
Comment (lukpueh) [inline tuf-spec.md:1344]:
```suggestion
 [[#update-root]] and version N of the root metadata file.
```

File: tuf-spec.md
Comment (lukpueh) [inline tuf-spec.md:1338]: There has also been a shift on the third level...
```suggestion
9. Repeat steps 5.3.2 to 5.3.9
```

File: tuf-spec.md
Comment (lukpueh) [inline tuf-spec.md:1491]:
```suggestion
     step [[#fetch-target]] (so that attackers cannot cause the client to waste excessive
```

File: tuf-spec.md
Comment (lukpueh) [inline tuf-spec.md:1567]:
```suggestion
versions. See step [[#update-root]]  of the client application workflow for more details.
```

File: tuf-spec.md
Comment (lukpueh) [inline tuf-spec.md:1646]:
```suggestion
See [[#detailed-client-workflow]] for more details.
```

**Review by lukpueh — CHANGES_REQUESTED** [review]
Here is the more thorough review as promised, which includes:
- a line by line review using `git diff --ignore-space-change --word-diff` (GitHub provides a [similar view](https://github.com/theupdateframework/specification/pull/143/files?w=1) with the `w=1` get param),
- a review of the those changes in the [rendered html](https://joshuagl.github.io/specification/),
- a grep on all words for which definitions exist (I added a few suggestions for additional references of already defined words, plus new definitions)
- a check of all cross-references to client workflow section (given that the numbering has shifted)

**Comment by trishankatdatadog** [comment]
> Here is the more thorough review as promised, which includes:

Spectacular review, thanks Lukas! 🙇🏽‍♂️

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:29]: Unfortunately that isn't supported by Bikeshed, which only [supports](https://tabatkins.github.io/bikeshed/#markdown) underlined and #-surrounded headings. I'd be happy to add a patch to switch to underlined headings, if you think that would be more readable?

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:523]: Agreed, will remove in a follow-on commit.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:527]: Nicely spotted!

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:741]: Good eye, I'll review the examples.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:674]: Good suggestion, I will add this.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:1009]: I had that at one point, but wasn't sure it made sense. I will re-add as you have also suggested it. We can improve the layout (it's a bit jarring jumping to the middle of a paragraph to find a definition) in future.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:1060]: Agreed, I an intending to spend some time working on the DELEGATION object definition in future.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:1022]: grammatically you can argue this one either way, but your suggestion seems to be the more common approach.

File: tuf-spec.md
Comment (lukpueh) [reply tuf-spec.md:29]: I think underlined headings only support two levels... at any rate, let's not overdo the bikeshedding. :)

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:4]: TS & DTS are only for documents that claim to be ISO process related documentation. Instead I'm defining a custom header that removes display of the Status text and lists the version right up at the top instead.

**Comment by joshuagl** [comment]
Thank you for the very detailed and high-quality review @lukpueh! I've addressed all of your suggestions for immediate work in follow-on commits, and added a few other minor changes. Most notably;
* added the extra cross-linking you suggested (7916026)
* integrated your grammatical clarifications (6708a28), removed "where:" (6671fb4), and tried to ensure definitions are consistently capitalised (df33652)
* fixed indentation in the examples (0c48395)
* added a custom HTML header to change the document heading – adding version and removing status (3fa355b)
* added a definition tag for the explanation of `KEY`s in the general principles section (3f5f5c0)
* cleaned up some unnecessary discussion of the "keys" field in `DELEGATION` objects (9c8305f)
* added a definition tag for `HEX_DIGEST` (2601e35)
* consolidated remarks about `KEYID`s in format principles (9d21a28)

**Comment by lukpueh** [comment]
High-quality work deserves a thorough review! :) Thanks for addressing my comments. I'll give it another pass on Monday.

**Comment by joshuagl** [comment]
A rendered version of the specification following the latest round of changes is viewable here: https://joshuagl.github.io/specification/1.0.17/

**Review by lukpueh — APPROVED** [review]

**Comment by lukpueh** [comment]
Absolutely amazing work, @joshuagl! Thanks a ton for taking care of this behemoth. Future readers of the spec should also thank you. Syncing changes into draft when we merge this might be a bit tricky, maybe we should do it in a PR so that someone can review it.

**Comment by joshuagl** [comment]
> Syncing changes into draft when we merge this might be a bit tricky, maybe we should do it in a PR so that someone can review it.

I've prepared a branch for that here https://github.com/joshuagl/specification/tree/joshuagl/bikeshed-initial-draft
Once this PR is merged I'll open another comparing that branch to the master branch so that we can review.

**Review by mnm678 — APPROVED** [review]
Thanks for the great work @joshuagl, and for the thorough review @lukpueh!

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:1267]: Filed as https://github.com/theupdateframework/specification/issues/146

**Pull request merged**
